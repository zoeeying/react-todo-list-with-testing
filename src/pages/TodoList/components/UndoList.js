import React, { Component } from 'react'
import '../index.css'

class UndoList extends Component {
  render() {
    const { list, deleteItem, changeStatus, onInputBlur, onInputChange } = this.props
    return (
      <div className="undo_list">
        <div className="undo_list_title">
          正在进行
          <div data-test="show-count" className="undo_list_count">
            {list.length}
          </div>
        </div>
        <ul className="undo_list_content">
          {list.map((item, index) => (
            <li
              data-test="list-item"
              key={index}
              className="undo_list_item"
              onClick={() => {
                changeStatus(index)
              }}
            >
              <span>
                {item.status === 'blur' ? (
                  item.value
                ) : (
                  <input
                    value={item.value}
                    className="list-input"
                    data-test="list-input"
                    onBlur={() => {
                      onInputBlur(index)
                    }}
                    onChange={e => {
                      onInputChange(index, e.target.value)
                    }}
                  ></input>
                )}
              </span>
              <div
                className="undo_list_delete"
                data-test="delete-item"
                onClick={e => {
                  e.stopPropagation()
                  deleteItem(index)
                }}
              >
                -
              </div>
            </li>
          ))}
        </ul>
      </div>
    )
  }
}

export default UndoList
