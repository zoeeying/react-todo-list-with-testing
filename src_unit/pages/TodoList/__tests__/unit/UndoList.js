import React from 'react'
import { shallow } from 'enzyme'
import UndoList from '../../components/UndoList'
import { findTestWrapper } from '@/utils/testUtils'

describe('测试UndoList组件', () => {
  const listData = [
    { value: '学习Jest', status: 'blur' },
    { value: '学习React', status: 'blur' },
    { value: '学习单元测试', status: 'blur' },
  ]

  it('UndoList组件样式渲染正常', () => {
    const wrapper = shallow(<UndoList list={[]} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('当传入数据为空数组时，count显示为0，列表无内容', () => {
    const wrapper = shallow(<UndoList list={[]} />)
    const countEle = findTestWrapper(wrapper, 'show-count')
    const listItems = findTestWrapper(wrapper, 'list-item')
    expect(countEle.text()).toBe('0')
    expect(listItems.length).toBe(0)
  })

  it('当传入数据不是空数组时，count显示为数组长度，列表有内容', () => {
    const wrapper = shallow(<UndoList list={listData} />)
    const countEle = findTestWrapper(wrapper, 'show-count')
    const listItems = findTestWrapper(wrapper, 'list-item')
    expect(countEle.text()).toBe('3')
    expect(listItems.length).toBe(3)
  })

  it('当传入数据不是空数组时，存在删除按钮', () => {
    const wrapper = shallow(<UndoList list={listData} />)
    const deleteItems = findTestWrapper(wrapper, 'delete-item')
    expect(deleteItems.length).toBe(3)
  })

  it('当传入数据不是空数组时，点击某个删除按钮，会调用删除方法', () => {
    const fn = jest.fn()
    const wrapper = shallow(<UndoList list={listData} deleteItem={fn} />)
    const deleteItems = findTestWrapper(wrapper, 'delete-item')
    // deleteItems是wrapper数组，如果想模拟点击第二个删除按钮
    // 不可以直接使用deleteItems[1]，需要使用deleteItems.at(1)
    const index = 1
    deleteItems.at(index).simulate('click', {
      stopPropagation: () => {},
    })
    expect(fn).toHaveBeenLastCalledWith(index) // fn最后一次被调用传入的参数是1
  })

  it('当点击某一项时，触发执行changeStatus函数', () => {
    const fn = jest.fn()
    const index = 1
    const wrapper = shallow(<UndoList list={listData} changeStatus={fn} />)
    const listItem = findTestWrapper(wrapper, 'list-item')
    listItem.at(index).simulate('click')
    expect(fn).toHaveBeenLastCalledWith(index) // fn最后一次被调用传入的参数是1
  })

  it('当某一项status是focus时，展示input输入框', () => {
    listData[0].status = 'focus'
    const wrapper = shallow(<UndoList list={listData} />)
    const inputItem = findTestWrapper(wrapper, 'list-input')
    expect(inputItem.length).toBe(1)
  })

  it('当某一个input输入框失去焦点时，触发onInputBlur函数', () => {
    listData[0].status = 'focus'
    const fn = jest.fn()
    const wrapper = shallow(<UndoList list={listData} onInputBlur={fn} />)
    const inputItem = findTestWrapper(wrapper, 'list-input')
    inputItem.simulate('blur')
    expect(fn).toHaveBeenLastCalledWith(0)
  })

  it('当某一个input输入框变更时，触发onInputChange函数', () => {
    listData[0].status = 'focus'
    const value = '小畅叙'
    const fn = jest.fn()
    const wrapper = shallow(<UndoList list={listData} onInputChange={fn} />)
    const inputItem = findTestWrapper(wrapper, 'list-input')
    inputItem.simulate('change', {
      target: { value },
    })
    expect(fn).toHaveBeenLastCalledWith(0, value)
  })
})
