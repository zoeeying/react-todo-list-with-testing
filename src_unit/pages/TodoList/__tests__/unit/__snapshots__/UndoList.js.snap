// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试UndoList组件 UndoList组件样式渲染正常 1`] = `
<div
  className="undo_list"
>
  <div
    className="undo_list_title"
  >
    正在进行
    <div
      className="undo_list_count"
      data-test="show-count"
    >
      0
    </div>
  </div>
  <ul
    className="undo_list_content"
  />
</div>
`;
